### ISSUE TITLE
Sorting weeks long recordings and/or merging subsorts

### ISSUE THREAD

**Issue by timsainb** [SpikeInterface/spikesorters#185]
Hi SpikeInterface team. Thanks for the awesome tool. 

Are there any protocols defined in spikeinterface for sorting days to weeks long recordings? 

I've played around with the different spike sorters and the only sorter that I've found works with >24 hour 32-channel recordings is IronSort. 

I looked into Kilosort/2/2.5, and [found a thread where @marius10p suggests splitting files and merging back neurons that look the same or overlap in time](https://github.com/MouseLand/Kilosort/issues/107#issuecomment-558210478).

I'm wondering if there are any protocols to e.g. merge spikes across subsorts, or if anyone has any code/experience doing this. 

My (naive) thoughts on how to do this now would be to split the recordings into e.g. 6 hr se.SubRecordingExtractor recordings, with say a 3-hour overlap. Then sort each 6 hr block individually. Then maybe write some protocol for merging based on a similarity of the spike template and the % overlap in spike times. That seems a hacky approach though. 

Does something like that already exist?

**Comment by mhhennig** [comment]
Interesting question!

So what you suggest may work: split into overlapping segments, and use the the spikecomparison tools (`spikecomparison.compare_two_sorters`) to find matching neurons in the overlapping segments. This should be quite easy to implement with the SubRecordingExtractors, as you suggest.

A main problem can be probe drift as it will not be compensated for. But it is worth a try I think. 

Herding Spikes can cope with large data, the largest recording (multiple files recorded from teh same prep) had >90 millon spikes. But HS was designed for dense arrays, I doubt it'll give good results for your data.

**Comment by timsainb** [comment]
Thanks @mhhennig finding overlapping units in the overlapping segments works very well. 

I'm currently at an impass on how to merge the overlapping sortings. Now that I have a set of putatively matching units, I want to:
1) merge together those two sortings into one SortingExtractor
2) merge overlapping units into a single unit

For the first part I can't find anything in the API that merges together the two sortings. Is there a good way of doing this with spikeinterface? I suppose the alternative would be to do this outside of spikeinterface and create a new se.NumpySortingExtractor. Then use SortingExtractor.copy_unit_properties and SortingExtractor.copy_unit_spike_features to copy features and properties over to the new numpy sorter. 

The second bit is straightforward, I can just call `spiketoolkit.curation.CurationSortingExtractor. merge_units`

**Comment by mhhennig** [comment]
You are right, merging sortings is not automated so far (this is the first use case...). It's also neccessary to correct the spike times for each segment so they are continuous, this should be done outside spikeinterface for now.

I wonder which functionality would help here. I supoose we could add a method to add spikes to an existing unit (@colehurwitz ?). But I expect there will be different numbers of units in each sorted segments, with some potentially unmatched (units may appear/disappear). It would be interesting to hear what you find there.

I myself have worked with some very large (1TB in total) high density recordings, and have concatenated them before sorting. Although it just about worked, this has its limitations as well...

**Comment by timsainb** [comment]
I used the numpy approach I mentioned above, and it works well. Like you said, units appear/disappear and so chaining together the units results in some overlapping clusters that are probably the same unit over time. 

I made a video of a subset spikes over the first 10 days (200K/90M), projected into UMAP (all channel activity for the 4ms surrounding the spike):
https://drive.google.com/file/d/1rMPbuVjHHRvd4TyWHe9JB5Wnoa0fBC7_/view?usp=sharing

The colors represent kilosort2 labels with the method mentioned above. I'll need to do some secondary curation to pull together clusters across days, but in principal it works.

**Comment by alejoe91** [comment]
@timsainb super!!!

Would you mind sharing the code with us? I think this would be a good addition for the API :)

**Comment by timsainb** [comment]
I merged them in a bit of a hacky way, but here's some pseudocode of what I did to merge two sorts in a chain:

```python
### to compare overlapping clusters
# get a subsort of the
sort_1_overlapping_part = se.SubSortingExtractor(
    parent_sorting=sort_1,
    start_frame=frame_where_overlapping_starts,
    end_frame=last_frame_in_sort_1,
)
sort_2_overlapping_part = se.SubSortingExtractor(
    parent_sorting=sort_2, start_frame=0, end_frame=frame_where_overlapping_ends,
)

# compare sortings
cmp_sort = sc.compare_two_sorters(
    sorting1=sort_1_overlapping_part,
    sorting2=sort_2_overlapping_part,
    sorting1_name="sort_1",
    sorting2_name="sort_2",
)
# grab overlapping units
overlapping_units = np.where((cmp_sort.agreement_scores > agreement_threshold).values)
prev_unit_label = list(cmp_sort.agreement_scores.index[overlapping_units[0]])
new_unit_label = list(cmp_sort.agreement_scores.columns[overlapping_units[1]])
max_sort_1_id = np.max(sort_1.get_unit_ids()) + 1
# update the unit IDs
updated_unit_ids = np.array(sort_2.get_unit_ids()) + max_sort_1_id
# update new unit labels
new_unit_label = np.array(new_unit_label) + max_sort_1_id

# get the part of the sort where spikes are overlapping
non_overlapping_addition = se.SubSortingExtractor(
    parent_sorting=sort_2,
    start_frame=frame_where_overlapping_ends,
    end_frame=last_frame_in_sort_2,
    renamed_unit_ids=updated_unit_ids,
)

# merge the sort_1 and non-overlapping segment
# get spikes and unit IDs for sort
sort_1_spikes = sort_1.get_units_spike_train()
sort_1_spike_units = [
    np.repeat(unit, len(spikes))
    for unit, spikes in zip(sort_1.get_unit_ids(), sort_1_spikes)
]
# get spikes and unit IDs for sort
new_spikes = [
    i + previous_sort_end for i in non_overlapping_addition.get_units_spike_train()
]
new_spike_units = [
    np.repeat(unit, len(spikes))
    for unit, spikes in zip(non_overlapping_addition.get_unit_ids(), new_spikes)
]
# grab spike times and labels for both sort_1 and new addition
merged_times = np.concatenate(
    [np.concatenate(sort_1_spikes), np.concatenate(new_spikes)]
)
merged_labels = np.concatenate(
    [np.concatenate(sort_1_spike_units), np.concatenate(new_spike_units),]
)
# create a numpy sorting extractor with new labels
merged_sorting = se.NumpySortingExtractor()
merged_sorting.set_times_labels(
    times=merged_times, labels=merged_labels,
)
merged_sorting.set_sampling_frequency(
    sampling_frequency=sort_1.get_sampling_frequency()
)
# we use this rather than get_unit_ids, in case some units arent in this part of the sort
non_overlapping_units = np.unique(np.concatenate(new_spike_units))
# copy unit properties
merged_sorting.copy_unit_properties(sort_1, unit_ids=sort_1.get_unit_ids())
merged_sorting.copy_unit_properties(
    non_overlapping_addition, unit_ids=non_overlapping_units
)

# copy spike_features
merged_sorting.copy_unit_spike_features(sort_1, unit_ids=sort_1.get_unit_ids())
merged_sorting.copy_unit_spike_features(
    non_overlapping_addition, unit_ids=non_overlapping_units
)

# merge each set of units above threshold
curated_merged_sorting = st.curation.CurationSortingExtractor(merged_sorting)
for prev_unit, new_unit in zip(prev_unit_label, new_unit_label):
    curated_merged_sorting.merge_units([prev_unit, new_unit])
```
